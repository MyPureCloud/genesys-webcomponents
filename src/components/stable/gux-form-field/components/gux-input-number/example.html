<form onchange="notify(event)" oninput="notify(event)">
  <h1>gux-form-field (Number)</h1>

  <p>slot a input[type=number] element into a gux-form-field element</p>

  <gux-form-field>
    <input
      slot="input"
      name="number"
      type="number"
      step="5"
      min="5"
      max="100"
      placeholder="Enter a number"
    />
    <label slot="label">Number</label>
  </gux-form-field>

  <gux-form-field clearable>
    <input
      slot="input"
      name="number-clearable"
      type="number"
      step="5"
      min="5"
      max="100"
      placeholder="Enter a number"
      value="50"
    />
    <label slot="label">Number</label>
  </gux-form-field>

  <gux-form-field>
    <input
      slot="input"
      name="number"
      type="number"
      step="5"
      min="5"
      max="100"
      placeholder="Enter a number"
      disabled
    />
    <label slot="label">Number</label>
  </gux-form-field>

  <gux-form-field clearable>
    <input
      slot="input"
      name="number-clearable"
      type="number"
      step="5"
      min="5"
      max="100"
      placeholder="Enter a number"
      value="50"
      disabled
    />
    <label slot="label">Number</label>
  </gux-form-field>
</form>
