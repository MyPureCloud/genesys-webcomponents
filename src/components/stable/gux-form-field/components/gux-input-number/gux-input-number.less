@import (reference) '../../../../../style/color.less';

// Variables part
@dark-background-color: @gux-type;
@dark-text-color: @gux-type-dark;
@dark-border: @gux-black-30;

gux-input-number {
  display: block;
  position: relative;

  .gux-input-number-container {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    justify-content: center;
    align-content: stretch;
    align-items: center;

    .gux-input-container {
      font-family: inherit;
      height: 32px;
      width: 100%;
      box-sizing: border-box;
      padding: 4px 12px 4px 12px;
      background-color: @gux-grey-100;
      background-image: none;
      border: 1px solid @gux-grey-60;
      border-radius: 2px;
      font-size: 12px;
      line-height: 1.66666667;
      color: @gux-black-50;
      display: flex;
      flex-direction: row;
      flex-wrap: nowrap;
      justify-content: center;
      align-content: stretch;
      align-items: center;
      order: 0;
      flex: 1 1 auto;
      align-self: auto;

      input {
        border: none;
        outline: none;
        order: 0;
        flex: 1 1 auto;
        align-self: auto;

        &::placeholder {
          color: @gux-black-90;
        }

        &::-webkit-input-placeholder {
          color: @gux-black-90;
        }

        &::-ms-input-placeholder {
          color: @gux-black-90;
        }

        &::-moz-placeholder {
          color: @gux-black-90;
        }

        &::-webkit-outer-spin-button,
        &::-webkit-inner-spin-button {
          -webkit-appearance: none;
          margin: 0;
        }

        /* Firefox */
        &[type='number'] {
          -moz-appearance: textfield;
        }
      }

      .gux-clear-button {
        order: 0;
        flex: 0 1 auto;
        align-self: auto;
        padding: 2px;
        border: none;
        background: transparent;
        cursor: pointer;
        color: @gux-black-90;

        &:focus,
        &:hover {
          color: @gux-blue-60;
        }

        gux-icon {
          width: 12px;
          height: 12px;
        }
      }
    }

    .gux-step-buttons-container {
      order: 0;
      flex: 0 1 16px;
      align-self: auto;
      margin: 0 4px;

      .gux-step-button {
        order: 0;
        flex: 0 1 auto;
        align-self: auto;
        padding: 2px;
        border: none;
        background: transparent;
        cursor: pointer;
        color: @gux-black-90;

        &:focus,
        &:hover {
          color: @gux-blue-60;
        }

        gux-icon {
          width: 10px;
          height: 10px;
        }
      }
    }

    &:focus-within {
      .gux-input-container {
        border: 1px solid @gux-blue-60;
        outline: none;
        box-shadow: 0 0 4px fade(@gux-blue-80, 50%);
      }
    }
  }
}

// Theming

// Dark
.gux-input-number-dark-theme {
  .gux-input-number-container {
    .gux-input-container {
      background-color: @dark-background-color;
      border: 1px solid @dark-border;
      color: @dark-text-color;

      &:focus-within {
        border: 1px solid @gux-blue-80;
        outline: 3px @gux-blue-80;
        box-shadow: none;
      }
    }
  }
}

.gux-dark-theme {
  gux-input-number {
    .gux-input-number-dark-theme();
  }
}

gux-input-number.gux-dark-theme {
  .gux-input-number-dark-theme();
}

// Light
.gux-input-number-light-theme {
  .gux-input-number-container {
    .gux-input-container {
      background-color: @gux-grey-100;
      border: 1px solid @gux-grey-60;
      color: @gux-black-50;
    }

    &:focus-within {
      .gux-input-container {
        border: 1px solid @gux-blue-60;
        box-shadow: 0 0 4px fade(@gux-blue-80, 50%);
      }
    }
  }
}

.gux-light-theme {
  gux-input-number {
    .gux-input-number-light-theme();
  }
}

gux-input-number.gux-light-theme {
  .gux-input-number-light-theme();
}

// Default Theme
gux-input-number {
  .gux-input-number-light-theme();
}
