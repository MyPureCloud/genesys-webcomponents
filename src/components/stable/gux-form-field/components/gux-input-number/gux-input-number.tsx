import { Component, Element, h, JSX, Prop, State } from '@stencil/core';

import { buildI18nForComponent, GetI18nValue } from '../../../../../i18n';
import setInputValue from '../../../../../utils/dom/set-input-value';
import simulateNativeEvent from '../../../../../utils/dom/simulate-native-event';

import modalComponentResources from './i18n/en.json';

/**
 * @slot input - Required slot for input[type="radio"]
 * @slot label - Required slot for label
 */
@Component({
  styleUrl: 'gux-input-number.less',
  tag: 'gux-input-number'
})
export class GuxInputNumber {
  private input: HTMLInputElement;
  private getI18nValue: GetI18nValue;

  @Element()
  private root: HTMLElement;

  @Prop()
  clearable: boolean;

  @State()
  private hasContent: boolean = false;

  private clearInput(): void {
    setInputValue(this.input, '', true);
  }

  private setHasContent(): void {
    this.hasContent = Boolean(this.input.value);
  }

  private simulateNativeInputAndChangeEvents(): void {
    simulateNativeEvent(this.input, 'input');
    simulateNativeEvent(this.input, 'change');
  }

  private stepUp(): void {
    if (this.input.value === '') {
      setInputValue(this.input, this.input.min || '0', false);
    } else {
      this.input.stepUp();
      this.simulateNativeInputAndChangeEvents();
    }
  }

  private stepDown(): void {
    if (this.input.value === '') {
      setInputValue(this.input, this.input.min || '0', false);
    } else {
      this.input.stepDown();
      this.simulateNativeInputAndChangeEvents();
    }
  }

  private renderClearButton(): JSX.Element {
    if (this.clearable && this.hasContent) {
      return (
        <button
          class="gux-clear-button"
          type="button"
          title={this.getI18nValue('clear')}
          onClick={this.clearInput.bind(this)}
        >
          <gux-icon iconName="ic-close" decorative></gux-icon>
        </button>
      );
    }

    return null;
  }

  private renderStepButtons(): JSX.Element {
    return (
      <div class="gux-step-buttons-container">
        <button
          class="gux-step-button"
          type="button"
          title={this.getI18nValue('increment')}
          onClick={() => this.stepUp()}
        >
          <gux-icon iconName="angle-up" decorative></gux-icon>
        </button>

        <button
          class="gux-step-button"
          type="button"
          title={this.getI18nValue('decrement')}
          onClick={() => this.stepDown()}
        >
          <gux-icon iconName="angle-down" decorative></gux-icon>
        </button>
      </div>
    );
  }

  async componentWillLoad(): Promise<void> {
    this.getI18nValue = await buildI18nForComponent(
      this.root,
      modalComponentResources
    );

    this.input = this.root.querySelector(
      'input[slot="input"], select[slot="input"]'
    );

    this.input.addEventListener('input', () => {
      this.setHasContent();
    });

    this.setHasContent();
  }

  render(): JSX.Element {
    return (
      <div class="gux-input-number-container">
        <div class="gux-input-container">
          <slot name="input" />
          {this.renderClearButton()}
        </div>
        {this.renderStepButtons()}
      </div>
    );
  }
}
