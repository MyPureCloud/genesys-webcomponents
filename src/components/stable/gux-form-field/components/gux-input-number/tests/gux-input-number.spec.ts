import { newSpecPage } from '@stencil/core/testing';
import { GuxInputNumber } from '../gux-input-number';

const components = [GuxInputNumber];
const language = 'en';

describe('gux-input-number', () => {
  it('should build', async () => {
    const html = `<gux-input-number><input slot="input" type="number"></gux-input-number>`;
    const page = await newSpecPage({ components, html, language });
    const component = page.rootInstance;

    expect(component).toBeInstanceOf(GuxInputNumber);
  });
});
