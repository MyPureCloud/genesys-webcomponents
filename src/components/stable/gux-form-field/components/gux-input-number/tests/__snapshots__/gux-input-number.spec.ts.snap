// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-input-number should render 1`] = `
<gux-input-number>
  <!---->
  <div class="gux-input-number-container">
    <div class="gux-input-container">
      <input slot="input" type="number">
    </div>
    <div class="gux-step-buttons-container">
      <button class="gux-step-button" title="Increment" type="button">
        <gux-icon decorative="" iconname="angle-up"></gux-icon>
      </button>
      <button class="gux-step-button" title="Decrement" type="button">
        <gux-icon decorative="" iconname="angle-down"></gux-icon>
      </button>
    </div>
  </div>
</gux-input-number>
`;
